algo/src/main/java/org/neo4j/graphalgo/impl/msbfs/MultiSourceBFS.java

```diff
@@ -0,0 +1,335 @@
+package org.neo4j.graphalgo.impl.msbfs;
+
+import org.neo4j.collection.primitive.PrimitiveIntIterator;
+import org.neo4j.graphalgo.api.IdMapping;
+import org.neo4j.graphalgo.api.RelationshipIterator;
+import org.neo4j.graphalgo.core.utils.ParallelUtil;
+import org.neo4j.graphdb.Direction;
+
+import java.util.Arrays;
+import java.util.concurrent.ExecutorService;
+
+/**
+ * Multi Source Breadth First Search implemented as described in [1].
+ * <p>
+ * The benefit of running this MS-BFS instead of multiple execution of a regular
+ * BFS for every source is that the MS-BFS algorithm can collapse traversals that are
+ * the same for multiple nodes. If any two or more given BFSs would traverse the same nodes
+ * at the same iteration depth, the MS-BFS will traverse only once and group all sources
+ * for this traversal.
+ * <p>
+ * The consumer of this algorithm provides a callback function, the gets called
+ * with:
+ * <ul>
+ * <li>the node id where the BFS traversal is at</li>
+ * <li>the depth or BFS iteration at which this node is traversed</li>
+ * <li>a lazily evaluated list of all source nodes that have arrived at this node at the same depth/iteration</li>
+ * </ul>
+ * The sources iterator is only valid during the execution of the callback and
+ * should not be stored.
+ * <p>
+ * We use a fixed {@code ω} (OMEGA) of 32, which allows us to implement the
+ * seen/visitNext bit sets as a packed long which improves memory locality
+ * as suggested in 4.1. of the paper.
+ * If the number of sources exceed 32, multiple instances of MS-BFS are run
+ * in parallel.
+ * <p>
+ * If the MS-BFS runs in parallel, the callback may be executed from multiple threads
+ * at the same time. The implementation should therefore be thread-safe.
+ * <p>
+ * The algorithm provides two invariants:
+ * <ul>
+ * <li>
+ * For a single thread, a single node is traversed at most once at a given depth
+ * – That is, the combination of {@code (nodeId, depth)} appears at most once per thread.
+ * It may be that a node is traversed multiple times, but then always at different depths.
+ * </li>
+ * <li>
+ * For multiple threads, the {@code (nodeId, depth)} may appear multiple times, but then always
+ * for different sources.
+ * </li>
+ * </ul>
+ * <p>
+ * [1]: <a href="http://www.vldb.org/pvldb/vol8/p449-then.pdf">The More the Merrier: Efficient Multi-Source Graph Traversal</a>
+ */
+public final class MultiSourceBFS implements Runnable {
+
+    // how many sources can be traversed simultaneously
+    private static final int OMEGA = 32;
+
+    private final ThreadLocal<MultiBitSet32> visits;
+    private final ThreadLocal<BiMultiBitSet32> nextAndSeens;
+
+    private final IdMapping nodeIds;
+    private final RelationshipIterator relationships;
```

Comment:
this indirection might prevent the JIT & CPU from performing certain optimizations, e.g., prefetching.
maybe it's not a big deal, but I'd consider revising the API. maybe prototype an implementation of this algorithm where all arrays are arrays (rather than hidden behind iterators) to see if it's really costing you anything.